### ISSUE TITLE
"Must have a LUN Index assigned" error for data disks...

### ISSUE THREAD

**Issue by ShamSV** [Azure/migAz#166]
I am using the latest version of the tool v2.4.2.2. I am getting the error "Virtual Machine 'xxxxxxxxxxx' Data Disk 'must have a LUN Index assigned.

The disk gets identified properly as Additional Resource under the VM, but when I go to that disk, there is no option to select the LUN ID.

Note: Screenshots attached.

The disk is a proper data disk, through Azure PowerShell.

HostCaching         : ReadWrite
DiskLabel           : 
DiskName            : XXXXXXXXXXXXXXX
Lun                 : 0
LogicalDiskSizeInGB : 1023
MediaLink           : https://XXXXXXXXXXXX.blob.core.windows.net/goldentemplate/xxxxxxxxxxxxxxx.vhd
SourceMediaLink     : 
IOType              : Standard
ExtensionData       : 

However in the portal, it shows under OS disk due to a know issue in the portal.

How can we get rid of the error and assign a LUN ID to this disk.

![index issue](https://user-images.githubusercontent.com/37436050/38493871-5bfc6538-3c11-11e8-87dd-052e35e9cb79.jpg)
![vm-tab](https://user-images.githubusercontent.com/37436050/38493876-5f19b2a2-3c11-11e8-9651-b20b23860d1e.jpg)

**Comment by rsanderson2350** [comment]
@ShamSV What type of disk is the source data disk?  i.e. ASM Disk, ARM Classic Disk, ARM Managed Disk?

**Comment by rsanderson2350** [comment]
@ctm-tech I'll have to review and consider the right option here.  How did you get a second "OS Disk" added to the VM?

**Comment by ctm-tech** [comment]
I inherited this configuration so have to admit that I do not know how they managed it

**Comment by ShamSV** [comment]
@rsanderson2350 : This is a ASM data disk. 
Even in my case, one of the data disk show as in the screenshot shared by @ctm-tech.

It is actually a data disk, but portal shows it like that.
PowerShell show the same disk as a valid data disk. I have cross verified it for few VMs.
This is a recently identified Azure Portal display issue. MS is already testing the fix for it.

**Comment by rsanderson2350** [comment]
@ctm-tech  I do not know how this was done either.  Is this VM an ASM VM, ARM VM with Classic Disks or ARM VM with Managed Disks?

Can you please locate the XML (If ASM) or JSON (if ARM) for this virtual machine (you'll need to locate it in the REST Responses)?  Please remove any sensitive information, such as the VM Name, Name of disks as you marked out in your screenshot, subscription id, etc.  I'll need to see the rest response for the definition of this VM to see what it is reporting for the OS and Data Disks.

**Comment by ShamSV** [comment]
Here is the disk section of the XML:
<DataVirtualHardDisks>
    <DataVirtualHardDisk>
      <HostCaching>ReadWrite</HostCaching>
      <DiskName>XXXXXXXXXXXXXXX-0-201510300535030544</DiskName>
      <Lun>0</Lun>
      <LogicalDiskSizeInGB>1023</LogicalDiskSizeInGB>
      <IOType>Standard</IOType>
    </DataVirtualHardDisk>
    <DataVirtualHardDisk>
      <HostCaching>None</HostCaching>
      <DiskName>XXXXXXXXXXXXXXX-1-201510300539350116</DiskName>
      <Lun>1</Lun>
      <LogicalDiskSizeInGB>1023</LogicalDiskSizeInGB>
      <IOType>Standard</IOType>
    </DataVirtualHardDisk>
    <DataVirtualHardDisk>
      <HostCaching>None</HostCaching>
      <DiskName>XXXXXXXXXXXXXXX-2-201510300541540131</DiskName>
      <Lun>2</Lun>
      <LogicalDiskSizeInGB>1023</LogicalDiskSizeInGB>
      <IOType>Standard</IOType>
    </DataVirtualHardDisk>
  </DataVirtualHardDisks>
  <OSVirtualHardDisk>
    <HostCaching>ReadWrite</HostCaching>
    <DiskName>cs-XXXXXXXXXXXXXXX-2015-10-30</DiskName>
    <SourceImageName>WIN2012-XXXXXXXXXXXXXXX</SourceImageName>
    <OS>Windows</OS>
    <IOType>Standard</IOType>
    <ResizedSizeInGB xsi:nil="true" />
  </OSVirtualHardDisk>

Its listing disk with LUN ID 0 appropriately as data disk.

**Comment by ctm-tech** [comment]
This is a Classic VM with Classic Disks,
XML Attached.
[XML.zip](https://github.com/Azure/migAz/files/1893825/XML.zip)

**Comment by ctm-tech** [comment]
Hi,
Just went to move another VM and have the same issue but this time the affected disk is under the Data disks section of the VM. This is an ASM > ARM migration. Sounds the same to @ShamSV

**Comment by ShamSV** [comment]
Hi,
The issue is consistent for ASM VM with data disks. The challenge is with the way disk 0 is handled.
The tool is not assigning LUN ID 0 to disk 0 (first data disk). In the previous versions, LUN ID option was not present in MigAZ tool. 
The portal issue of showing the data disk as OS disk is sorted. It shows the disks appropriately.

**Comment by rsanderson2350** [comment]
I have not been able to reproduce the issue in which the Source ASM VM is not getting / passing the Data Disk LUNs into the Target definition.  I did a test also using the XML provided, and all Data Disks has their respective LUN index pull across.

I have added a code update to safe guard Data Disks in which the LUN can't be determined (source of this problem has not been identified yet), and if the LUN Index can't be obtained for a Data Disk, it will be defaulted to an invalid index of -1.  In this case, you will be required to set the LUN Index (which should be set the same as the source).
